compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs

```diff
@@ -120,36 +138,29 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                     && proj_base.is_empty()
                     && !self.upvars.is_empty()
                 {
-                    item_msg = access_place_desc;
                     debug_assert!(self.body.local_decls[ty::CAPTURE_STRUCT_LOCAL].ty.is_ref());
                     debug_assert!(is_closure_or_coroutine(
                         the_place_err.ty(self.body, self.infcx.tcx).ty
                     ));
 
-                    reason = if self.is_upvar_field_projection(access_place.as_ref()).is_some() {
-                        ", as it is a captured variable in a `Fn` closure".to_string()
+                    diagnostic = if self.is_upvar_field_projection(access_place.as_ref()).is_some()
+                    {
+                        PlaceAndReason::SelfCaptured(access_place_desc)
                     } else {
-                        ", as `Fn` closures cannot mutate their captured variables".to_string()
+                        PlaceAndReason::UpvarCaptured(access_place_desc)
                     }
                 } else {
-                    let source = self.borrowed_content_source(PlaceRef {
+                    let ptr_source = self.borrowed_content_source(PlaceRef {
                         local: the_place_err.local,
                         projection: proj_base,
                     });
-                    let pointer_type = source.describe_for_immutable_place(self.infcx.tcx);
-                    opt_source = Some(source);
-                    if let Some(desc) = self.describe_place(access_place.as_ref()) {
-                        item_msg = format!("`{desc}`");
-                        reason = match error_access {
-                            AccessKind::Mutate => format!(", which is behind {pointer_type}"),
-                            AccessKind::MutableBorrow => {
-                                format!(", as it is behind {pointer_type}")
-                            }
-                        }
-                    } else {
-                        item_msg = format!("data in {pointer_type}");
-                        reason = String::new();
-                    }
+                    let name = ptr_source.describe_for_immutable_place(self.infcx.tcx);
+                    diagnostic = PlaceAndReason::BehindPointer(
+                        self.describe_place_typed(access_place.as_ref()),
```

Comment:
Ah, so the "value" case in `DescribedPlace` is to handle the situation of "data" here?